components/eamxx/src/diagnostics/aerocom_cld.cpp

```diff
@@ -0,0 +1,208 @@
+  using namespace ekat::units;
+  using namespace ShortFieldTagsNames;
+
+  auto grid             = grids_manager->get_grid("Physics");
+  const auto &grid_name = grid->name();
+
+  const auto nondim = Units::nondimensional();
+  const auto micron = m / 1000000;
+
+  m_ncols = grid->get_num_local_dofs();
+  m_nlevs = grid->get_num_vertical_levels();
+  m_ndiag = 8;
+
+  // Define layouts we need (both inputs and outputs)
+  FieldLayout scalar2d_layout{{COL, LEV}, {m_ncols, m_nlevs}};
+  FieldLayout vector1d_layout{{COL, CMP}, {m_ncols, m_ndiag}};
+
+  // The fields required for this diagnostic to be computed
+  add_field<Required>("T_mid", scalar2d_layout, K, grid_name);
+  add_field<Required>("pseudo_density", scalar2d_layout, Pa, grid_name);
+  add_field<Required>("p_mid", scalar2d_layout, Pa, grid_name);
+  add_field<Required>("qv", scalar2d_layout, kg / kg, grid_name);
+  add_field<Required>("qc", scalar2d_layout, kg / kg, grid_name);
+  add_field<Required>("qi", scalar2d_layout, kg / kg, grid_name);
+  add_field<Required>("eff_radius_qc", scalar2d_layout, micron, grid_name);
+  add_field<Required>("eff_radius_qi", scalar2d_layout, micron, grid_name);
+  add_field<Required>("cldfrac_tot", scalar2d_layout, nondim, grid_name);
+  add_field<Required>("nc", scalar2d_layout, kg / kg, grid_name);
+
+  // Construct and allocate the aodvis field
+  FieldIdentifier fid(name(), vector1d_layout, nondim, grid_name);
+  m_diagnostic_output = Field(fid);
+  m_diagnostic_output.allocate_view();
+
+  // Self-document the outputs to parse in post-processing
+  using stratt_t = std::map<std::string, std::string>;
+  auto d         = get_diagnostic();
+  auto &metadata =
+      d.get_header().get_extra_data<stratt_t>("io: string attributes");
+  metadata["0"] = "T_mid";
+  metadata["1"] = "p_mid";
+  metadata["2"] = "cldfrac_ice";
+  metadata["3"] = "cldfrac_liq";
+  metadata["4"] = "cdnc";
+  metadata["5"] = "eff_radius_qc";
+  metadata["6"] = "eff_radius_qi";
+  metadata["7"] = "cldfrac_tot";
+}
+
+void AeroComCld::compute_diagnostic_impl() {
+  /* The goal of this routine/impl is to calculate properties at cloud top
+   * based on the AeroCom recommendation. See reference for routine
+   * get_subcolumn_mask in rrtmpg, where equation 14 is used for the
+   * maximum-random overlap assumption for subcolumn generation. We use
+   * equation 13, the column counterpart. The logic is reversed to calculate
+   * cloud-bottom properties as well.
+   */
+  using KT  = KokkosTypes<DefaultDevice>;
+  using MT  = typename KT::MemberType;
+  using ESU = ekat::ExeSpaceUtils<typename KT::ExeSpace>;
+
+  using PF = scream::PhysicsFunctions<DefaultDevice>;
+
+  std::vector<int> level_vector(m_nlevs);
+  for(int i = 1; i < m_nlevs; ++i) {
+    level_vector[i] = m_topbot == "Top" ? i : m_nlevs - i;
+  }
+
+  const auto out = m_diagnostic_output.get_view<Real **>();
+
+  // zero out the outputs
+  Kokkos::deep_copy(out, 0.0);
```

Comment:
Just a personal preference, but I would move lines 90-98 directly before the `Kokkos::parallel_for()` since that is the only place that uses these.